src/zernike_fet.cpp

```diff
@@ -0,0 +1,291 @@
+  tally_shape.insert(tally_shape.end(), cylinder_shape.begin(),
+                     cylinder_shape.end());
+
+  // another dimnesion of length two, which is related to the zernike
+  // responsible for radial and legendre reposible for the axis. first will be
+  // zernike and second will be legendre.
+  tally_shape.push_back(2);
+
+  // last-dimension is for the orders, however, the radial will incorporated by
+  // zernike and axial will be legendre, therefore, for both the order can be
+  // same or different, therefore, a max of both will be taken, so all the tally
+  // will remain at one place.
+  std::size_t max_order = zr_order_;
+  if (legen_order_ > zr_order_) {
+    max_order = legen_order_;
+  }
+  tally_shape.push_back(max_order + 1);
+
+  // reallocate and fill with zeros for the taly avg, gen-score and varaince
+  tally_avg_.reallocate(tally_shape);
+  tally_avg_.fill(0.0);
+
+  tally_gen_score_.reallocate(tally_shape);
+  tally_gen_score_.fill(0.0);
+
+  tally_var_.reallocate(tally_shape);
+  tally_var_.fill(0.0);
+
+  // construct the zernike polynomials upto the given order
+  zr_polynomial_ = ZernikePolynomials(zr_order_);
+}
+
+void ZernikeFET::score_collision(const Particle& p, const Tracker& tktr,
+                                 MaterialHelper& mat) {
+  StaticVector6 indices;
+  // get the energy-index, if energy-filter exists
+  if (energy_filter_) {
+    std::optional<std::size_t> E_indx = energy_filter_->get_index(p.E());
+
+    if (E_indx.has_value()) {
+      std::size_t index_E = E_indx.value();
+      indices.push_back(index_E);
+    } else
+      // Not inside any energy bin. Don't score.
+      return;
+  }
+
+  // get the positional indexes from cylinder_filter
+  StaticVector3 cylinder_index = cylinder_filter_->get_indices(tktr);
+  if (cylinder_index.empty()) {
+    // Not inside any energy bin. Don't score.
+    return;
+  }
+  indices.insert(indices.begin(), cylinder_index.begin(), cylinder_index.end());
+
+  // add one dimension for the different polynomial
+  std::size_t poly_index = indices.size();
+  indices.push_back(0);
+
+  // add one dimension for the different order
+  std::size_t FET_index = indices.size();
+  indices.push_back(0);
+
+  // get the particle base score for the collision
+  const double Et = mat.Et(p.E());
+  const double collision_score = particle_base_score(p, mat) / Et;
+
+  // varaible for scoring
+  double beta_n;
+
+  // first- score for the zernike
+  std::pair<double, double> scaled_r_and_theta =
+      cylinder_filter_->get_scaled_radius_and_angle(cylinder_index, tktr.r());
+  const double scaled_r = scaled_r_and_theta.first;
+  const double theta = scaled_r_and_theta.second;
+  const std::vector<double> zr_value =
+      zr_polynomial_.evaluate_zernikes(scaled_r, theta);
+  if (zr_value.size() != (zr_order_ + 1)) {
+    fatal_error(
+        "for the zernike-FET, something happened as the sizes for evaluated "
+        "polynomials and zr-order+1 is not same.");
+  }
```

Comment:
We probably don't need this check here. We can be very certain this will not happen, and adding this check will slow down tallying.